core/src/main/scala/org/apache/spark/scheduler/BlacklistTracker.scala

```diff
@@ -0,0 +1,214 @@
+    val now = clock.getTimeMillis()
+    // quickly check if we've got anything to expire from blacklist -- if not, avoid doing any work
+    if (now > nextExpiryTime) {
+      val execsToClear = executorIdToBlacklistStatus.filter(_._2.expiryTime < now).keys
+      if (execsToClear.nonEmpty) {
+        logInfo(s"Removing executors $execsToClear from blacklist during periodic recovery")
+        execsToClear.foreach { exec =>
+          val status = executorIdToBlacklistStatus.remove(exec).get
+          val failedExecsOnNode = nodeToFailedExecs(status.node)
+          failedExecsOnNode.remove(exec)
+          if (failedExecsOnNode.isEmpty) {
+            nodeToFailedExecs.remove(status.node)
+          }
+        }
+      }
+      if (executorIdToBlacklistStatus.nonEmpty) {
+        nextExpiryTime = executorIdToBlacklistStatus.map{_._2.expiryTime}.min
+      } else {
+        nextExpiryTime = Long.MaxValue
+      }
+      val nodesToClear = nodeIdToBlacklistExpiryTime.filter(_._2 < now).keys
+      if (nodesToClear.nonEmpty) {
+        logInfo(s"Removing nodes $nodesToClear from blacklist during periodic recovery")
+        nodesToClear.foreach { node => nodeIdToBlacklistExpiryTime.remove(node) }
+        // make a copy of the blacklisted nodes so nodeBlacklist() is threadsafe
+        _nodeBlacklist.set(nodeIdToBlacklistExpiryTime.keySet.toSet)
+      }
+    }
+ }
+
+  def taskSetSucceeded(failuresByExec: HashMap[String, FailureStatus]): Unit = {
+    // if any tasks failed, we count them towards the overall failure count for the executor at
+    // this point.
+    failuresByExec.foreach { case (exec, newFailures) =>
+      val prevFailures = executorIdToFailureCount.getOrElse(exec, 0)
+      val newTotal = prevFailures + newFailures.totalFailures
+      val node = newFailures.node
+
+      if (newTotal >= MAX_FAILURES_PER_EXEC) {
+        logInfo(s"Blacklisting executor id: $exec because it has $newTotal" +
+          s" task failures in successful task sets")
+        val now = clock.getTimeMillis()
+        val expiryTime = now + EXECUTOR_RECOVERY_MILLIS
+        executorIdToBlacklistStatus.put(exec, BlacklistedExecutor(node, expiryTime))
+        val blacklistedExecsOnNode = nodeToFailedExecs.getOrElseUpdate(node, HashSet[String]())
+        blacklistedExecsOnNode += exec
+        executorIdToFailureCount.remove(exec)
+        if (expiryTime < nextExpiryTime) {
+          nextExpiryTime = expiryTime
+        }
+
+        if (blacklistedExecsOnNode.size >= MAX_FAILED_EXEC_PER_NODE) {
+          logInfo(s"Blacklisting node $node because it has ${blacklistedExecsOnNode.size} " +
+            s"executors blacklisted: ${blacklistedExecsOnNode}")
+          nodeIdToBlacklistExpiryTime.put(node, expiryTime)
+          // make a copy of the blacklisted nodes so nodeBlacklist() is threadsafe
+          _nodeBlacklist.set(nodeIdToBlacklistExpiryTime.keySet.toSet)
+        }
+      } else {
+        executorIdToFailureCount.put(exec, newTotal)
```

Comment:
For readability maybe just move this to line 102? It's harmless (and negligible extra time) in the case when the executor needs to be blacklisted since then it will be removed.